### DISCUSSION TITLE
Queue

### DISCUSSION THREAD

**Discussion by BLoHny** [subak-ba/cs-study#33] [자료 구조]
# Queue
![image](https://github.com/subak-ba/cs-study/assets/113517233/34e9bd92-f661-4b53-89c7-1d703fa8073e)
<br>

## 정의
- 선입선출(FIFO)의 자료구조<br><br>

## 주요 연산
- push(item) : 큐 맨뒤에 데이터 삽입 하는 연산
  시간복잡도 O(1)
![image](https://github.com/subak-ba/cs-study/assets/113517233/c040903a-11b4-4a20-b99d-1b22b364f9c6)

- pop() : 큐 맨앞 데이터 삭제 및 반환 하는 연산
  시간복잡도 O(1)
![image](https://github.com/subak-ba/cs-study/assets/113517233/63019f7d-f929-454d-883c-b49eba802c3f)

- peek() : 현재 front 위치의 데이터를 반환만 하는 연산

![image](https://github.com/subak-ba/cs-study/assets/113517233/3cd4bb94-92fe-44f1-a51d-aa23c7eb9101)

<br>

## If Queue is Full
- 예외 처리 -> ex ) IllegalStateException
- 자동 확장 -> 기존 값 유지후, 큐 크기 증가
- 오버라이트 -> 가장 오래된(FIrst In)값을 오버라이트 : 추천되지않음 👎 
- 블로킹 큐 -> Multi-Threading 환경에서 동기화를 위한 방법

<br>

## 주요 베리에이션

- ### 원형 큐(환형 큐)
  
![image](https://github.com/subak-ba/cs-study/assets/113517233/5ef865fe-6fd1-41bb-89f2-235321e07987)

배열 길이가 한정되있고, 확장이 없는 기준에서, 기존의 큐는 공간이 꽉 차게 되면 더 이상 요소를 추가할 수 없었습니다. 심지어 앞쪽에 요소들이 pop()을 통하여 제거되더라도 front와 tail의 위치가 같아져 front와 tail을 0 으로 돌리기 전까지는 앞쪽이 메모리를 낭비하는 현상또한 있었습니다.

원형 큐에서는 그것을 해결하기 위해 배열의 맨 마지막 부분을 쓰면 다시 맨 첫 부분부터 사용하여 낭비를 최소화 합니다.
  
- ### 우선순위 큐
  
![image](https://github.com/subak-ba/cs-study/assets/113517233/ee942404-428b-499a-a254-0ffef4569818)

우선순위 큐란 말그대로 각 원소들에게 우선순위를 매겨 First In 과 상관없이 요소를 pop 시키는 큐입니다.
예시로는 Heap 이있습니다.

힙과 우선순위 큐의 차이점 -> 우선순위 큐는 ADT 힙은 구현체일뿐이다.

- ### 데큐
  
![image](https://github.com/subak-ba/cs-study/assets/113517233/bc612489-34e6-4009-b27b-1ec221f145be)

일반 적인 큐가 뒤에서 삽입 앞에서 인출이 가능한 반면, 데큐에서는 어느 방향에서나 데이터 삽입 인출이 가능합니다.

<br>

## 구현 코드 예시(size 확장)
```java
import java.io.IOException;
import java.util.Arrays;

public class Main<T> {
    private int capacity;
    private int size;
    private int front;
    private int rear;
    private T[] queue;

    public Main() {
        capacity = 5;
        size = 0;
        front = 0;
        rear = -1;
        queue = (T[]) new Object[capacity];
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void enqueue(T item) {
        if (isEmpty()) {
            front = 0;
            rear = -1;
        }

        if (size == capacity) {
            resize(2 * capacity);
        }
        rear = (rear + 1) % capacity;
        queue[rear] = item;
        size++;
    }

    public void dequeue() {
        if (isEmpty()) {
            throw new IllegalStateException("Queue is empty");
        }
        T item = queue[front];
        queue[front] = null; // 아이템 삭제
        front = (front + 1) % capacity;
        size--;
        System.out.println(item);
    }


    public int size() {
        return size;
    }

    private void resize(int newCapacity) {
        T[] newQueue = (T[]) new Object[newCapacity];
        for (int i = 0; i < size; i++) {
            newQueue[i] = queue[(front + i) % capacity];
        }
        queue = newQueue;
        front = 0;
        rear = size - 1;
        capacity = newCapacity;
    }

    public T peek() {
        if (isEmpty()) {
            throw new IllegalStateException("Queue is empty");
        }
        return queue[front];
    }

    public static void main(String[] args) throws IOException {
        Main<Integer> main = new Main<Integer>();

        main.enqueue(1);
        main.enqueue(2);
        main.enqueue(3);
        main.enqueue(4);
        main.enqueue(5);
        main.enqueue(6);

        System.out.println(Arrays.toString(main.queue));

        main.dequeue();
        main.dequeue();
        main.dequeue();
        main.dequeue();
        main.dequeue();
        main.dequeue();

        System.out.println(Arrays.toString(main.queue));

        main.enqueue(7);

        System.out.println(Arrays.toString(main.queue));
    }
}
```
**출력 결과**
```cmd
[1, 2, 3, 4, 5, 6, null, null, null, null]
1
2
3
4
5
6
[null, null, null, null, null, null, null, null, null, null]
[7, null, null, null, null, null, null, null, null, null]
```

<br>

## 문제
10845:큐 - https://www.acmicpc.net/problem/10845 (실버)

**Comment by onee-only** [comment]
블로킹 큐가 어떤 동작인지 설명이 모호한 것 같아요. 보충 설명을 추가 해 주실 수 있나요?
